oidb/pepper/service/src/jwk.rs

```diff
@@ -0,0 +1,91 @@
+use crate::{Issuer, KeyID};
+use anyhow::{anyhow, Result};
+use dashmap::DashMap;
+use jsonwebtoken::{
+    jwk::{Jwk, JwkSet},
+    DecodingKey,
+};
+use log::{info, warn};
+use once_cell::sync::Lazy;
+use std::{sync::Arc, thread::sleep, time::Duration};
+
+/// The JWK in-mem cache.
+pub static DECODING_KEY_CACHE: Lazy<DashMap<Issuer, DashMap<KeyID, Arc<DecodingKey>>>> =
+    Lazy::new(DashMap::new);
+
+/// Send a request to a JWK endpoint and return its JWK map.
+pub async fn fetch_jwks(jwk_url: &str) -> Result<DashMap<KeyID, Arc<DecodingKey>>> {
+    let response = reqwest::get(jwk_url)
+        .await
+        .map_err(|e| anyhow!("jwk fetch error: {}", e))?;
+    let text = response
+        .text()
+        .await
+        .map_err(|e| anyhow!("error while getting response as text: {}", e))?;
+    let JwkSet { keys } = serde_json::from_str(text.as_str())
+        .map_err(|e| anyhow!("error while parsing json: {}", e))?;
+    let key_map: DashMap<KeyID, Arc<DecodingKey>> = keys
+        .into_iter()
+        .filter_map(
+            |jwk| match (&jwk.common.key_id, DecodingKey::from_jwk(&jwk)) {
+                (Some(kid), Ok(key)) => Some((kid.clone(), Arc::new(key))),
+                (Some(kid), Err(e)) => {
+                    warn!("error while parsing for kid {kid}: {e}");
+                    None
+                },
+                (None, _) => {
+                    warn!("Ignoring a kid-less jwk: {jwk:?}");
+                    None
+                },
+            },
+        )
+        .collect();
+    Ok(key_map)
+}
+
+pub fn start_jwk_refresh_loop(issuer: &str, jwk_url: &str, refresh_interval: Duration) {
+    let issuer = issuer.to_string();
+    let jwk_url = jwk_url.to_string();
+    let _handle = tokio::spawn(async move {
+        loop {
+            match fetch_jwks(jwk_url.as_str()).await {
+                Ok(key_set) => {
+                    let num_keys = key_set.len();
+                    DECODING_KEY_CACHE.insert(issuer.clone(), key_set);
+                    info!(
+                        "Updated key set of issuer {}. Num of keys: {}.",
+                        issuer, num_keys
+                    );
+                },
+                Err(msg) => {
+                    warn!("{}", msg);
+                },
+            }
+            sleep(refresh_interval);
+        }
+    });
+}
+
+pub fn cached_decoding_key(issuer: &String, kid: &String) -> Result<Arc<DecodingKey>> {
```

Comment:
will this function be changed?

Reply:
in which way? getting rid of the test jwk?

Reply:
yeah, we have both test and non-test code in the same "prod" function, which behaves differently based on user input. what are the consequences if in a production flow someones specifies "test_jwk" in their claims?

Reply:
this will be deleted. we can test with real jwks. cc @heliuchuan